WcaOnRails/app/javascript/edit-events/EditEvents.jsx

```diff
@@ -0,0 +1,245 @@
+import React from 'react'
+import cn from 'classnames'
+import ReactDOM from 'react-dom'
+
+import events from 'wca/events.js.erb'
+import { rootRender, promiseSaveWcif } from 'edit-events'
+import { EditTimeLimitButton, EditCutoffButton, EditAdvancementConditionButton } from 'edit-events/modals'
+
+export default class EditEvents extends React.Component {
+  constructor(props) {
+    super(props);
+    this.save = this.save.bind(this);
+    this.onUnload = this.onUnload.bind(this);
+  }
+
+  save(e) {
+    let {competitionId, wcifEvents} = this.props;
+    let wcif = {
+      id: competitionId,
+      events: wcifEvents,
+    };
+
+    this.setState({ saving: true });
+    promiseSaveWcif(wcif).then(response => {
+      return response.json().then(json => [ response, json ]);
+    }).then(([response, json]) => {
+      if(!response.ok) {
+        throw new Error(`${response.status}: ${response.statusText}\n${json["error"]}`);
+      }
+      this.setState({ savedWcifEvents: clone(this.props.wcifEvents), saving: false });
+    }).catch(e => {
+      this.setState({ saving: false });
+      alert("Something went wrong while saving.\n" + e.message);
+    });
+  }
+
+  unsavedChanges() {
+    return !deepEqual(this.state.savedWcifEvents, this.props.wcifEvents);
+  }
+
+  onUnload(e) {
+    if(this.unsavedChanges()) {
+      var confirmationMessage = "\o/";
+      e.returnValue = confirmationMessage;
+      return confirmationMessage;
+    }
+  }
```

Comment:
Ah, yikes, this looks hacky. Please see https://developer.mozilla.org/en-US/docs/Web/API/WindowEventHandlers/onbeforeunload. I've also updated the code to hopefully make it more clear what's going on here. It's kind of silly because there's this `confirmationMessage` that gets ignored by every browser.